### ISSUE TITLE
Shorter Review Iterations

### ISSUE THREAD

**Issue by sils** [coala/coala#4080]
Hey people. I feel like we're spending a lot of time on review and not too much time on getting actual stuff done. (Also due to lots of newcomer issues.) Working on improving that is an ongoing process.

I do think though that we are a bit strict in the reviews in some places, e.g. commit messages, which shifts the focus from good code to good metadata and much frustration.

@coala/coala-developers @coala/coala-maintainers what do you think about this? Can we afford to be less strict in some areas and keep up our quality?

**Comment by meetmangukiya** [comment]
We can limit the number of reviewers on a PR. This will reduce the review requests and hence the shorter iterations. 
But here we hold the assigned reviewers responsible for getting the PR in a decent state at least.
Also we might loose on some good reviews from other reviewers who might've reviewed but didn't review because we limited the reviewers. To reduce this, we have to find appropriate reviewers for the PR. 

For selecting appropriate reviewer we can do:
1. Maintain a mapping of each file with persons(maintainers/developers only) who know about that particular file in the repo
    1. One way would be the authors of the file.
    2. Other way would be to manually find them.
2. Randomly select a couple of people from the mapping of the files being changed and assign as reviewers.

But, this is a whole new workflow, but I don't know if we want this :P 
To setup a reliable mapping system can take significant amount of time.

**Comment by meetmangukiya** [comment]
> One way would be the authors of the file.

I've seen similar thing being done in the `cURL`  repo

**Comment by ashwini0529** [comment]
Specially for New Comers, it becomes so hard to change commit messages again and again, and to review them is also bit tedious. I think we can afford to be relaxed atleast for Newcomers. But we have to keep in mind that the quality of the commit message etc doesn't get lost.

**Comment by arjunsinghy96** [comment]
For commit messages, we could emphasize only on the short log and the issue link.

**Comment by vijeth-aradhya** [comment]
I feel review processes should be strict as they are, but we could reduce the time we put in for each PR per day I guess, and work on other improvements side by side.

- One of the best things about coala I have seen is merging *strongly* reviewed code. The quality should remain the same.

And, about commits and other stuff, the newcomers can just read the guidelines, and we could help them a couple of times and they'll get used to it after one PR or so. We could relax things a bit I guess.

- **Also, I think we should keep a discussion session every two weeks where in we can discuss about new features, improvements, plans and specific troubling issues.**

**Comment by Mariatta** [comment]
I agree with making things less strict to contributors, and especially to newcomer and first time contributors.
Maintainers have the ability to edit the pull request. If a contributor's pull request look good, but require minor adjustment, (e.g. spacing, updating commit message), the maintainer can do the edit themselves instead of asking the contributor to do it. I would recommend that maintainers choose to make these adjustments themselves especially if the PR is made by a newcomer or first time contributors.

**Comment by Mixih** [comment]
The slight problem with maintainers edits is that it makes the branch diverge from the pr maker's tree. This can make it harder for them to edit their branch should the need arise. (or is this actually a problem....?)

**Comment by Mariatta** [comment]
The edit from maintainer should be done at the last step. Once maintainer make the edit, it should be in a 'ready to commit' stage. So no further update is needed in that branch, so I don't see any problem.
If further updates is needed, no reason not to first accept the PR and make another PR to address it?

**Comment by NiklasMM** [comment]
I'd argue against making the review less strict. 

When I first joined the strict, but transparent requirements for PRs and thorough, but helpful review process was one of the things that immediately impressed me. I think *especially* for newcomers who might never have worked on collaborative project or might be unfamiliar with git, it's important that they learn how things are done here. 

One thing that we could improve is doing more complete reviews. Often a reviewer will find a couple of typos and mark the PR as wip, without taking a closer look. Then a day goes by until the typos are fixed, before the next reviewer comes by. I think if someone reviews a PR, they should completely review it, in order to limit iterations.

**Comment by Mariatta** [comment]
True it is important for newcomers to learn how things are done, but they don't have to learn everything all at once. This is something that they can learn over time, they don't have to have a perfect PR the first time.

When we make our process too strict, for some people, it can become a barrier and an intimidating situation, rather than being welcoming.

In my mind, making the process less strict does not equate lowering our standards.
Maintainers still have the ability to adjust a PR. I expect the maintainers to know what our requirements are, and how to adjust the PR so that it is of high quality.

When a maintainer adjusts the PR, they should still explain what was changed, so that PR author can learn from it, and we hope that their subsequent PR will become better.

When we make their first experience awesome and they felt welcome, it will motivate them to stay and make another contribution.

In their next contribution, we can let them know that these are things that need change, and then we request them to make those changes. If after 2 iterations, maintainers still think that there are further changes, in this case I would suggest the maintainer to make these changes themselves, and then accept the PR. Also, we may need to think why a PR need more than 2 iterations before it can be merged. Perhaps we weren't being clear about what changes do we want to see?

**Comment by NiklasMM** [comment]
The barrier and intimidation aspect is real, but I think it does not come from the amount of things that are criticized, but how the criticism is conveyed. If we say things like:

> Your commit message must be changed like this, because of that. (Read here for more info)
> And consider making the following adjustments to your code, because of this (Link to background info)

instead of: 

> You're commit message sucks and I don't thing your code is valid.

That can make all the difference. 

I also think that making the PR merge ready after it is opened is an important part of the process and the newcomer issues are mostly typo fixing exactly *because* people can focus on the "meta".

**Comment by Makman2** [comment]
There are some criterias for me when reviewing a PR:
- Commit metadata
  - Shortlog
  - issue reference
  - Body: I'm not so strict about commit bodies if the issue reference is good itself
- Commit subdivision
   If something is atomic, put it inside an own commit. Easier to review + better history.
- Code
  - Layout (readable code = well reviewable code = better understandable code -> better code)
    this means often some spacing issues, alignments
  - Logic
    - compare how I would do it with the code flow
    - if code uses `i = 0` or other indexes, it's very likely it can be made shorter and better and faster (in fact if anywhere numbers occur, someone should take a closer look)
    - Code clones
    - Look if breakdown of stuff into functions/classes/other units is possible
    - ...
  - Tests (not so strict about them too, as long as they cover all newly added code, it's only a benefit)

What now is the juicy part is the "Code" part, especially the Logic. Depending on the complexity of the problem, reviewing time raises exponentially. Also reviewing logic is absolutely a matter of concentration. Sometimes you are more concentrated, sometimes not. Sometimes you see something sometimes not. That's why I often still find issues after my first reviews.

To maintain the same quality, I can't spend less time on looking at logic (or generally the actual code), that's a fact I believe, as this *is* what we actually want to achieve when solving an issue or generally improving something in coala. It just has to be good to work properly and nicely. I.e.: there I can't see a way lowering the number of review iterations (except if the PR submitter makes better code, which is a matter of training).

So what's remaining are "Commit metadata" and "Commit subdivision". Those are actually pretty easy once you get the hang of them, so we may be able to optimize there. I think our standards are good as they are, we developed them with all experiences we got during the coala project. That means

- we have to teach the people better to care for the proper "packaging" of their code.
  People are self-responsible for doing right commits and metadata. They need to read up on stuff themselves and we criticize if something's missing when they submit commits. This means that developers or maintainers do have to take some time to write motivating, yet precise comments. Or
- we have to do it for them. 
  Improves speed, but lessens the developer's learning experience. This also means that developers or maintainers will have to spend time to correct commits (this is not that much of time for commit message changes, but still)

For the first point: Maybe more examples inside documentations do help. Or we have to write a really strict guideline, in the form "1. Do this 2. Do commit exactly like this 3. Press Enter". There's a possibility to make videos, but I do not like videos at all, they are too time consuming and boring imo ;)

For the second point: No problem for me to change someone's PR. Though if someone would just change one my PRs without a good reason, I would actually be a bit mad. Tell me, I do the changes, so I can better keep track of what's happening with my work. We may ask people if they want corrections from our side.
That aside, there's still a technical limitation: I can't modify PRs in-place if they are done from a fork. I would need to ask those people to give me access -> some people don't know how to do that -> then I need to explain that. And then maybe telling them "please do `git commit --amend` and change msg to xxx followed by a `git push --force`" is shorter. Or I close their PR and make a new one, though I dislike that because the submitter can't modify the PR any more (although doing this would be the final step until merge).

That's how things are things are for me, I know this is no real answer to the problem, though maybe some considerable viewpoint ;)

**Comment by satwikkansal** [comment]
My viewpoint: I think the newcomer labeled issues are a good way to introduce the people to the workflow of coala (including the strict guidelines for code and commit). The only concern imo is sometimes it may get frustrating for the newcomer to make several iterations for a single line change, eventually making them to give up. The main reasons imo for large number of iterations of a newcomer PR are:
- Multiple reviewers asking the author to change the commit message but giving just hints instead of exact solution. This may result in newcomer updating commit-message again and again due to new requested changes in every iteration (which is very frustrating).
- Reviewers marking a PR WIP on the basis of commit message or semantical changes/typos. This removes the PR from the review queue until the author makes another iteration by addressing those minor suggested changes. 
- Too many rebase but the PR not actually not getting merged because of either the author is not able to update the PR at the right time or the maintainer is not available (after the PR is rebased)
 
The idea of @NiklasMM of giving a comprehensive review (instead of asking them to just to go and read the guide) seems nice to me. 

To reduce the number of iterations, few ideas that come to my mind:
1. Reviewing the code functionality should be the foremost priority of the reviewer in the initial iterations. We need to avoid superficial reviews and encourage in-depth quality reviews.
2. A reviewer should request for commit message change only after all the code looks functionally correct to him.
3. Only the maintainer should ask for rebase and that too when he/she is immediately going to merge the PR. (Maybe we can add a "only-rebase-needed" label to the PR)
4.  The PR should be marked WIP only when significant functionality changes have been requested in the code. Not marking WIP will invite other reviewers to suggest their changes and would also leave a scope for discussion among other reviewers before the author finally implements the changes.(Maybe the author can himself mark the PR wip, once he feels he has gathered enough inputs to work on his next iteration. Or if there's no activity on the PR for a week or so and the requested changes are still to be addressed, then any reviewer can mark the PR WIP.)
5. We can have a separate newcomer queue for newcomer PRs (as sims suggested), but I'm not sure if we should let other newcomers review these PRs, rather we should encourage them to review low difficulty PRs because it will also reflect their capability of being part of the developer team. 

Also, another idea can be having a "final-iteration" label for PRs which only need:
- commit changes (message change, or squash)
- Minor semantical changes
- Rebase
This way, when a PR with "final-iteration" label is pushed, a maintainer can proofread the PR and merge it immediately.

**Comment by Makman2** [comment]
> Multiple reviewers asking the author to change the commit message but giving just hints instead of exact solution. This may result in newcomer updating commit-message again and again due to new requested changes in every iteration (which is very frustrating).

I think this is one of the main points we should consider^^

**Comment by jayvdb** [comment]
There are a few related GSoC projects happening this year regarding newcomer process, and also reviewing the reviewers.  http://github.com/coala/community and http://github.com/coala/gh-board (and similar) will become our replacements for the review queues, and can be improved to provide extra functionality to assist reviewing.

Also this year, 

- Fixes/Closes commit message validation, avoiding divergent (often outdated) interpretations of the rule
- an automated commit message suggester, will provide newcomers with a recommended commit message for small changes, and hopefully that will prevent reviewers giving divergent suggestions to newcomers.
